go-controller/pkg/ovn/controller/admin_network_policy/admin_network_policy.go

```diff
@@ -32,6 +49,888 @@ func (c *Controller) processNextANPWorkItem(wg *sync.WaitGroup) bool {
+// gressRule cache object for that Rule.
+func cloneANPIngressRule(raw anpapi.AdminNetworkPolicyIngressRule, priority int32) (*gressRule, error) {
+	anpRule := &gressRule{
+		name:        raw.Name,
+		priority:    priority,
+		action:      getACLActionForANPRule(raw.Action),
+		gressPrefix: ANPIngressPrefix,
+		peers:       make([]*adminNetworkPolicyPeer, 0),
+		ports:       make([]*adminNetworkPolicyPort, 0),
+	}
+	for _, peer := range raw.From {
+		anpPeer, err := cloneANPPeer(peer)
+		if err != nil {
+			return nil, err
+		}
+		anpRule.peers = append(anpRule.peers, anpPeer)
+	}
+	if raw.Ports != nil {
+		for _, port := range *raw.Ports {
+			anpPort := cloneANPPort(port)
+			anpRule.ports = append(anpRule.ports, anpPort)
+		}
+	}
+
+	return anpRule, nil
+}
+
+// cloneANPEgressRule takes the provided ANP API Egres Rule and creates a new corresponding
+// gressRule cache object for that Rule.
+func cloneANPEgressRule(raw anpapi.AdminNetworkPolicyEgressRule, priority int32) (*gressRule, error) {
+	anpRule := &gressRule{
+		name:        raw.Name,
+		priority:    priority,
+		action:      getACLActionForANPRule(raw.Action),
+		gressPrefix: ANPEgressPrefix,
+		peers:       make([]*adminNetworkPolicyPeer, 0),
+		ports:       make([]*adminNetworkPolicyPort, 0),
+	}
+	for _, peer := range raw.To {
+		anpPeer, err := cloneANPPeer(peer)
+		if err != nil {
+			return nil, err
+		}
+		anpRule.peers = append(anpRule.peers, anpPeer)
+	}
+	if raw.Ports != nil {
+		for _, port := range *raw.Ports {
+			anpPort := cloneANPPort(port)
+			anpRule.ports = append(anpRule.ports, anpPort)
+		}
+	}
+	return anpRule, nil
+}
+
+// syncAdminNetworkPolicy decides the main logic everytime
+// we dequeue a key from the anpQueue cache
 func (c *Controller) syncAdminNetworkPolicy(key string) error {
+	// TODO(tssurya): This global lock will be inefficient, we will do perf runs and improve if needed
+	c.Lock()
+	defer c.Unlock()
+	startTime := time.Now()
+	_, anpName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	klog.V(4).Infof("Processing sync for Admin Network Policy %s", anpName)
+
+	defer func() {
+		klog.V(4).Infof("Finished syncing Admin Network Policy %s : %v", anpName, time.Since(startTime))
+	}()
+
+	anp, err := c.anpLister.Get(anpName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+	if anp == nil {
+		// it was deleted; let's clear up all the related resources to that
+		err = c.clearAdminNetworkPolicy(anpName)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	// at this stage the ANP exists in the cluster
+	return c.setAdminNetworkPolicy(anp)
+}
+
+// setAdminNetworkPolicy will handle the main reconcile logic for any given anp's
+// add/update that might be triggered either due to ANP changes or the corresponding
+// matching pod or namespace changes.
+func (c *Controller) setAdminNetworkPolicy(anp *anpapi.AdminNetworkPolicy) error {
```

Comment:
> Honestly, this method seems overly complicated. The only reasonable optimization (on a first take) I see is:
> 
> * If anything in the ANP itself changes, send all the ops as if we were adding the policy from scratch
>   
>   * Whenever we add a policy from scratch, since we don't know if we are starting up and that policy already existed and we did actually missed an update, we just delete all the pre-exisistng AS for the policy by predicate.
> * Otherwise, if new pods need to be considered as subject or peers, just send the update ops for the PG/Address sets
> 
> As it has been mentioned elsewhere in this PR, ANPs should not be updating as much. And as we have also done elsewhere in this PR, I think its's reasonable to take the simple approach first and optimize later if needed.

So this is what I started with in predecessor PR: https://github.com/ovn-org/ovn-kubernetes/pull/3489 where I would just delete all the stuff and recreate it based on current API state - doesn't matter which field in the ANP changed, but then we had a design change for a central loop controller. So now every pod/namespace change that happens, we will queue the ANP key back to the main control loop which means each pod/namespace change will call this function - so then I started doing all these perms/combs and that is how we ended up here. If we feel this is complicated and should just go back to delete/create always that's fine but I have no way to know or differentiate whether it was a pod/namespace change or an ANP change...

However I have not done scale runs yet (scheduled for next release).

> 
> Also think must of the logic to build the appropriate ops could be offloaded to separate method so that this method focuses on a higher level, calling a single entry point to build of ops, then transact and use/update the caches appropriately; would help with clarity.

ack lemme see if I can clean this up better..

> 
> Would rename to `ensureAdminNetworkPolicy` which is the common naming scheme we use in other places.

yup I am fine with calling this `ensureAdminNetworkPolicy` -> this is done, but let me leave the comment  open for more dialogue next week on which way we want to model this sync.

Reply:
> So this is what I started with in predecessor PR: #3489 where I would just delete all the stuff and recreate it based on current API state - doesn't matter which field in the ANP changed, but then we had a design change for a central loop controller. So now every pod/namespace change that happens, we will queue the ANP key back to the main control loop which means each pod/namespace change will call this function - so then I started doing all these perms/combs and that is how we ended up here. If we feel this is complicated and should just go back to delete/create always that's fine but I have no way to know or differentiate whether it was a pod/namespace change or an ANP change...

Yeah, that's why I was contemplating basically handling two different things: a full sync or a specific AS/PG sync. But this seemed more complicated than that to me.